File: lib/models/device_session.dart
class DeviceSession {
  final String sessionId;
  final String deviceId;
  final String deviceName;
  final String platform;
  final DateTime createdAt;
  final DateTime lastSeen;
  final bool isActive;

  const DeviceSession({
    required this.sessionId,
    required this.deviceId,
    required this.deviceName,
    required this.platform,
    required this.createdAt,
    required this.lastSeen,
    required this.isActive,
  });

  factory DeviceSession.fromJson(Map<String, dynamic> json) {
    return DeviceSession(
      sessionId: json['session_id'] as String,
      deviceId: json['device_id'] as String,
      deviceName: json['device_name'] as String,
      platform: json['platform'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      lastSeen: DateTime.parse(json['last_seen'] as String),
      isActive: json['is_active'] as bool,
    );
  }
}

File: backend/app/services/device_session.py
from datetime import datetime, timezone

from sqlalchemy import delete, select
from sqlalchemy.orm import Session

from app.models.device_session import DeviceSession
from app.schemas.device_session import DeviceSessionCreate


def get_user_device_sessions(
    db: Session,
    user_id: int,
) -> list[DeviceSession]:
    return list(
        db.scalars(
            select(DeviceSession)
            .where(DeviceSession.user_id == user_id)
            .order_by(DeviceSession.last_seen.desc())
        )
    )


def create_or_update_device_session(
    db: Session,
    user_id: int,
    device_data: DeviceSessionCreate,
) -> DeviceSession:
    device = db.scalar(
        select(DeviceSession)
        .where(
            DeviceSession.user_id == user_id,
            DeviceSession.device_id == device_data.device_id,
        )
    )

    now = datetime.now(timezone.utc)

    if device is None:
        device = DeviceSession(
            user_id=user_id,
            device_id=device_data.device_id,
            device_name=device_data.device_name,
            platform=device_data.platform,
            created_at=now,
            last_seen=now,
            is_active=True,
        )

        db.add(device)

    else:
        device.device_name = device_data.device_name
        device.platform = device_data.platform
        device.last_seen = now
        device.is_active = True

    db.commit()
    db.refresh(device)

    return device


def update_device_last_seen(
    db: Session,
    session_id: str,
) -> DeviceSession | None:
    device = db.scalar(
        select(DeviceSession)
        .where(DeviceSession.session_id == session_id)
    )

    if device is None:
        return None

    device.last_seen = datetime.now(timezone.utc)
    device.is_active = True

    db.commit()
    db.refresh(device)

    return device


def deactivate_device_session(
    db: Session,
    user_id: int,
    session_id: str,
) -> bool:
    device = db.scalar(
        select(DeviceSession)
        .where(
            DeviceSession.session_id == session_id,
            DeviceSession.user_id == user_id,
        )
    )

    if device is None:
        return False

    device.is_active = False

    db.commit()

    return True


def delete_device_session(
    db: Session,
    user_id: int,
    session_id: str,
) -> bool:
    result = db.execute(
        delete(DeviceSession).where(
            DeviceSession.session_id == session_id,
            DeviceSession.user_id == user_id,
        )
    )

    db.commit()

    return result.rowcount > 0